### PR TITLE
feat: introduce Slider

### PR DESCRIPTION
**Pull request by IlyaSurmay**
## Related Issue
Closes SAP/fundamental-styles#1709

## Description
Introduce Slider styles

## Screenshots

image
image
image

#### Please check whether the PR fulfills the following requirements

1. The output matches the design specs
- [x] All values are in `rem`
- [x] Text elements follow the truncation rules
- [x] hover state of the element follow design spec
- [x] focus state of the element follow design spec
- [x] active state of the element follow design spec
- [x] selected state of the element follow design spec
- [x] selected hover state of the element follow design spec
- [x] pressed state of the element follow design spec
- [x] Responsiveness rules - the component has modifier classes for all breakpoints
- [x] Includes Compact/Cosy/Tablet design
- [x] RTL support
2. The code follows fundamental-styles code standards and style
- [x] BEM naming convention is used
- [x] Mixins are used for repeatable code (`fd-rtl`, `fd-ellipsis`, `fd-flex`, `fd-selected`, `fd-focus`, ect.)
- [x] A11y support - keyboard support, screenreader support, proper ARIA attributes, etc.
- [x] `fd-reset()` mixin is applied to all elements
- [x] Variables are used, if some value is used more than twice.
- [x] Checked if current components can be reused, instead of having new code.
3. Testing
- [x] tested Storybook examples with "CSS Resources" `normalize` option 
- [x] tested Storybook examples with "CSS Resources" `unnormalize` option 
- [x] Verified all styles in IE11
- [x] Updated tests
4. Documentation
- [x] Storybook documentation has been created/updated
- [x] Breaking Changes [wiki](https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes) has been updated in case of breaking changes.

### CODE REVIEW

**Comment by droshev** [comment]
the numbers don't look centered with the slider tracks
Screen Shot 2020-10-08 at 3 34 23 PM

File: src/slider.scss
Comment (InnaAtanasova) [inline src/slider.scss:34]: this block can't be here in the way you use it. If it's an element (like you used it) it should be outside of the `&__track` block since it's a different element.
If you want to keep it here you need to make it a modifier class and rename it to `--range`. In this case you will have to use both classes on your div: 
```
<div class="fd-slider__track fd-slider__track--range"></div>
```

In this case inside the `&--track` block you don't need the reset, the position: relative, border-radius, etc., as they will come from `fd-slider__track ` class. That's why you make a modifier class, to modify stuff from the base class. Which approach you will choose is up to you :)

File: src/slider.scss
Comment (InnaAtanasova) [inline src/slider.scss:57]: there are mixins for the flex. Check `fd-flex()`

File: src/slider.scss
Comment (InnaAtanasova) [inline src/slider.scss:64]: You can check the flex mixins in the mixins file

File: src/slider.scss
Comment (InnaAtanasova) [inline src/slider.scss:93]: can you put these in variables?

File: src/slider.scss
Comment (InnaAtanasova) [inline src/slider.scss:106]: can you put these in variables?

File: src/slider.scss
Comment (InnaAtanasova) [inline src/slider.scss:124]: these should be in variables

File: src/slider.scss
Comment (InnaAtanasova) [inline src/slider.scss:128]: same as above

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]

File: stories/slider/slider.stories.js
Comment (InnaAtanasova) [inline stories/slider/slider.stories.js:40]: why do you add inline styling here? Our components should have inline styling only for better representation in the documentation (usually it's applied on the container div).

File: src/theming/sap_fiori_3.scss
Comment (jbadan) [inline src/theming/sap_fiori_3.scss:48]: Custom css variables are meant for situations where high contrast has a different value than light/dark and sap-theming repo has not defined a variable for it. A lot of these seem to be the same in all themes? If so, just use the `var(--sap*)` variable in `slider.scss`

For example, ` fdSlider_Track_Background`, `fdSlider_Handle_Background`, `fdSlider_Handle_Border_Color`, `fdSlider_Handle_Background_Hover`, `fdSlider_Handle_Border_Color_Hover`, `fdSlider_Handle_Background_Active`, `fdSlider_Handle_Border_Color_Active` should be removed and just use the sap variable in their place in the code. 

The goal is to keep these theming files as small as possible as we slowly work these values back into sap theming repo - we ideally want to have 0 custom css variables in this repository.

File: stories/slider/slider.stories.js
Comment (jbadan) [reply stories/slider/slider.stories.js:40]: I feel like the inline css might be ok _for this component only_. It's being used to display the different states of the slider without having to go through the trouble of setting up the interaction to work.

We should take the `<div style="width: 50%">` off the wrapper and use one of the docs classes instead though. You can find some containers here - https://github.com/SAP/fundamental-styles/blob/master/.storybook/custom/custom.scss or add another if none of those fit the needs of this component.

File: stories/slider/slider.stories.js
Comment (meganaconley) [reply stories/slider/slider.stories.js:35]: The slider in the example is also an input. It's possible to have it be a div, but we'll need a manual tab-index as well. Here's an example of a slider made up of divs: https://www.w3.org/TR/wai-aria-practices/examples/slider/slider-1.html

Specifically we need the aria-valuemin/valuemax/valuenow attributes, the tabindex, and an aria-label.

I have not been able to find an example of a range slider that uses a div, however; they all use an input.

**Review by InnaAtanasova — COMMENTED** [review]
Removing the requested changes as I am going on vacation

File: stories/slider/slider.stories.js
Comment (IlyaSurmay) [reply stories/slider/slider.stories.js:40]: @jbadan that's right, inline styling on slider handles was used to give them different positions so that way user sees a more or less real-world example of how range slider would look like. Of course, in component implementation these values are going to be set dynamically.

I'll use helper class from custom scss file to set wrapper's width, thanks

**Comment by IlyaSurmay** [comment]
> the numbers don't look centered with the slider tracks
> Screen Shot 2020-10-08 at 3 34 23 PM

that's right, they aren't aligned perfectly but I don't think this needs (and can be) done with CSS only because labels position depends on many factors like what text is used, label's width, how many labels there are
I believe that each label position should be calculated on the component level in JS implementation as CSS doesn't provide enough possibilities to do that, please correct me if I'm wrong. Also, Fiori design guidelines mentions that labels should have a responsive mechanism which will be activated if there is not enough horizontal space to display all the labels.

For demo purposes I think I can tweak CSS or add inline styling to align labels with ticks if that's necessary.

**Comment by jbadan** [comment]
@droshev We might need to take a look at the CLA bot - did it really take _4 days_ to check this pr?

**Review by jbadan — CHANGES_REQUESTED** [review]
Requesting changes until a11y feedback is resolved

File: src/theming/sap_fiori_3.scss
Comment (IlyaSurmay) [reply src/theming/sap_fiori_3.scss]: thanks! I removed variables that are the same for all themes

File: src/slider.scss
Comment (JKMarkowski) [inline src/slider.scss:54]: Let's change it to `top: 50%` and `transform: translateY(-50%);` for proper center

File: src/slider.scss
Comment (JKMarkowski) [inline src/slider.scss:103]: You don't need those values. center it by `top: 50%` and `transform: translate(-50%, -50%)`;

File: src/slider.scss
Comment (JKMarkowski) [inline src/slider.scss:128]: Please change `fd-screen` to `--lg` modifier

File: src/slider.scss
Comment (JKMarkowski) [inline src/slider.scss:129]: Also remove it, because it's going to be centered anyway

File: src/slider.scss
Comment (JKMarkowski) [inline src/slider.scss:91]: Let's add fd-reset there

**Review by JKMarkowski — CHANGES_REQUESTED** [review]
Hello @IlyaSurmay Great Job! I left some comment regarding mostly centering elements. Use `left/top` + `translate` to make sure that element will be always centered.
Also mobile mode example is missing there.

File: src/slider.scss
Comment (IlyaSurmay) [reply src/slider.scss:58]: this will only work when tick labels aren't displayed, that's why I added just `top` here

File: stories/slider/slider.stories.js
Comment (dimamarksman) [inline stories/slider/slider.stories.js:49]: Maybe it should be `fd-slider__inner` since it lays in `fd-slider` and doesn't wrap it?

File: src/slider.scss
Comment (dimamarksman) [inline src/slider.scss:100]: Can we center labels doing this? 
image

**Review by dimamarksman — COMMENTED** [review]
Looks good!
Couple of questions.

File: src/slider.scss
Comment (IlyaSurmay) [reply src/slider.scss:100]: awesome solution, thank you!

File: stories/slider/slider.stories.js
Comment (jacobdevera) [inline stories/slider/slider.stories.js:40]: Similar to the single-handle examples, the handles themselves should be the ones with role `slider` with the associated ARIA attributes as they would be treated as two separate tabbable elements. [Example](https://www.w3.org/TR/wai-aria-practices/examples/slider/multithumb-slider.html)

```suggestion
    <div class="fd-slider">
        <div class="fd-slider__inner">
            <div class="fd-slider__track">
                <div class="fd-slider__track-range" style="width: 40%;left: 40%"></div>
                <div class="fd-slider__handle" tabindex="0" role="slider" aria-label="min slider" aria-valuemin="1" aria-valuemax="80" aria-valuenow="40" style="left: 40%;"></div>
                <div class="fd-slider__handle" tabindex="0" role="slider" aria-label="max slider" aria-valuemin="40" aria-valuemax="100" aria-valuenow="80" style="left: 80%;"></div>
            </div>
```

File: stories/slider/slider.stories.js
Comment (jacobdevera) [inline stories/slider/slider.stories.js:24]: The handle itself should contain the `slider` attributes as it will be the element to receive focus.

```suggestion
    <div class="fd-slider">
        <div class="fd-slider__inner">
            <div class="fd-slider__track">
                <div class="fd-slider__track-range" style="width: 50%;"></div>
                <div class="fd-slider__handle" tabindex="0" role="slider" aria-label="slider" aria-valuemin="1" aria-valuemax="100" aria-valuenow="50" style="left: 50%;"></div>
            </div>
        </div>
    </div>
```

It's fine that the role is not on the main `fd-slider` element; notice in [this example](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_slider_role#Example_2_Text_Values) how the "slider" class is used as an outer container but the handle receives the actual slider role.

**Review by jacobdevera — CHANGES_REQUESTED** [review]
Few adjustments to where slider attributes should be placed. Thanks for taking accessibility into consideration!

File: stories/slider/slider.stories.js
Comment (IlyaSurmay) [reply stories/slider/slider.stories.js]: Thanks! I moved ARIA attributes to handle elements.

However, it seems to me that it's different for other libraries, for instance, Angular Material Slider has ARIA attributes on the host container and ng5-slider on handle. Anyway, let's stick to official guidelines from MDN.

**Review by JKMarkowski — APPROVED** [review]
LGTM

**Review by jbadan — APPROVED** [review]
Looks great to me (pending sign off from a11y team member)! 🚢

**Review by jacobdevera — APPROVED** [review]
LGTM 👍

File: stories/slider/slider.stories.js
Comment (jacobdevera) [reply stories/slider/slider.stories.js]: It's definitely possible to keep the attributes on the host container given appropriate CSS and proper focus management, though it's a bit harder, especially with multiple handles.

**Review by prsdthkr — CHANGES_REQUESTED** [review]
I'm unable to see the focus indicator when tabbing through the stories.

File: stories/slider/slider.visual.js
Comment (jbadan) [inline stories/slider/slider.visual.js:1]: We just moved to automated visual testing via chromatic 🎉 , so you can remove this file
Docs: https://github.com/SAP/fundamental-styles/wiki/Visual-testing-with-Chromatic
PR: https://github.com/SAP/fundamental-styles/pull/1584

Comment (IlyaSurmay) [reply stories/slider/slider.visual.js]: Cool :) 
removed

**Comment by prsdthkr** [comment]
Should focus align to this slider example seen in the spec here: https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?pageId=2046198993 ?

image

**Comment by jacobdevera** [comment]
> Should focus align to this slider example seen in the spec here: https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?pageId=2046198993 ?
If we do it this way, wouldn't it be impossible to manage focus between multi-handle sliders?

**Comment by prsdthkr** [comment]
> 
> > Should focus align to this slider example seen in the spec here: https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?pageId=2046198993 ?
> 
> If we do it this way, wouldn't it be impossible to manage focus between multi-handle sliders?

Good point.

On the handle, it should align to the standard Fiori focus outline on all zoom levels.
image

I'm seeing a different focus indicator on zoom level 125 in Chrome 🤔 :
image

**Comment by IlyaSurmay** [comment]
> > > Should focus align to this slider example seen in the spec here: https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?pageId=2046198993 ?
> > 
> > 
> > If we do it this way, wouldn't it be impossible to manage focus between multi-handle sliders?
> 
> Good point.
> 
> On the handle, it should align to the standard Fiori focus outline on all zoom levels.
> image
> 
> I'm seeing a different focus indicator on zoom level 125 in Chrome 🤔 :
> image

I couldn't reproduce this in latest Chrome with zoom 125%, outline looks exactly the same
Снимок экрана 2020-10-16 в 12 53 49

**Comment by prsdthkr** [comment]
@IlyaSurmay Cool, I think it is just me.

Also, don't we need an example for Element F and G as per spec here: https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?pageId=2033852394 ?
image

**Comment by IlyaSurmay** [comment]
> @IlyaSurmay Cool, I think it is just me.
> 
> Also, don't we need an example for Element F and G as per spec here: https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?pageId=2033852394 ?
> image

@prsdthkr 
I think examples with popover and input aren't necessary in this case because they aren't related to slider's styles, they're more like an integration of different Fiori components. Also, It'd mess with slider's markup in examples and make it harder to understand because each implementation (ng, vue, react) will use its own way of including popover/input in slider. 

And one more thing is that including markup of other components in slider is adding additional complexity because if there are any breaking changes in popover/input markup we'll have to change it in sliders' examples as well.

**Comment by prsdthkr** [comment]
I think it is okay to compose styles from various components to achieve the spec. We do it with many fd-styles components. It reduces ambiguity and standardizes the implementations. However, I'll defer to @droshev and team to take the call on this.

**Comment by prsdthkr** [comment]
Also, Range (Right to Left) seems to be misstyled: https://5f85e3904b19d600221872af-lfemkaaqsw.chromatic.com/?path=/story/visual-light-slider--slider

image

**Comment by IlyaSurmay** [comment]
> Also, Range (Right to Left) seems to be misstyled: https://5f85e3904b19d600221872af-lfemkaaqsw.chromatic.com/?path=/story/visual-light-slider--slider
> 
> image

@prsdthkr 
That's an interesting one. Selected range element is positioned with using `left` property and for basic slider it works fine in both LTR/RTL modes but for range slider the value of `left` needs to be changed, I'm not sure this can be achieved with CSS only because the offset of selected range element has to be recalculated based on the position of handles. Of course this won't be a big issue in JS implementations but in CSS it is.

Any ideas on this, am I missing something?

**Review by JKMarkowski — APPROVED** [review]
LGTM, great job

**Comment by IlyaSurmay** [comment]
@prsdthkr 
I fixed this issue with wrong offset of selected range element (for this demo only), please take a look

**Comment by JKMarkowski** [comment]
> > Also, Range (Right to Left) seems to be misstyled: https://5f85e3904b19d600221872af-lfemkaaqsw.chromatic.com/?path=/story/visual-light-slider--slider
> > image
> 
> @prsdthkr
> That's an interesting one. Selected range element is positioned using `left` property and for basic slider it works fine in both LTR/RTL modes but for range slider the value of `left` needs to be changed, I'm not sure this can be achieved with CSS only because the offset of selected range element has to be recalculated based on the position of handles. Of course this won't be a big issue in JS implementations but in CSS it is.
> 
> Any ideas on this, am I missing something?

The thing you can do there is to add class with left: 40% and width: 40%. And change `left` to `-20%`, when there is RTL

File: stories/slider/slider.stories.js
Comment (meganaconley) [inline stories/slider/slider.stories.js:21]: After looking into this some more, this should really be an `input`. Using `div`s will put a large burden on the implementation libraries to handle keyboard and screen reader support. Mobile screen readers in particular seem to struggle with this. Using an `input` would make the experience consistent across all platforms and assistive technologies.

See the following examples:
https://butterpep.com/slider-accessibility.html
https://thepaciellogroup.github.io/AT-browser-tests/test-files/input-range.html
https://www.w3.org/TR/wai-aria-1.1/#slider

**Review by meganaconley — CHANGES_REQUESTED** [review]

File: stories/slider/slider.stories.js
Comment (IlyaSurmay) [reply stories/slider/slider.stories.js:21]: > After looking into this some more, this should really be an `input`. Using `div`s will put a large burden on the implementation libraries to handle keyboard and screen reader support. Mobile screen readers in particular seem to struggle with this. Using an `input` would make the experience consistent across all platforms and assistive technologies.
> 
> See the following examples:
> https://butterpep.com/slider-accessibility.html
> https://thepaciellogroup.github.io/AT-browser-tests/test-files/input-range.html
> https://www.w3.org/TR/wai-aria-1.1/#slider

I got your point. However, I think that using `div` is OK in this case for a variety of reasons:
- poor styling abilities for `input[type="range"]` and inability to style in according to Fiori spec without both CSS/JS hacks
- implementing a range slider on native `input[type="range"]` will be way more complicated compared to `div`s
- handling keyboard/screen reader support and adding required `ARIA` attributes doesn't seem to be a large burden to me because frameworks provide in-built support of such functionality (Angular/React/Vue)
- existing implementation of an old Fiori slider is using `div`s (https://sapui5.hana.ondemand.com/1.82.4/#/entity/sap.m.RangeSlider/sample/sap.m.sample.RangeSlider)
- in addition, I couldn't find a popular 3rd party slider lib for Angular/React/Vue that uses `input[type="range"]`, IMO this is because of the first three reasons

> - https://material.angular.io/components/slider/overview
> - https://angular-slider.github.io/ng5-slider/
> - https://www.telerik.com/kendo-angular-ui/components/inputs/slider/
> - https://www.primefaces.org/primeng/showcase/#/slider
> - https://material-ui.com/components/slider/
> - https://zillow.github.io/react-slider/

Comment (meganaconley) [reply stories/slider/slider.stories.js:21]: I'm not sure that other 3rd party libraries not doing this is necessarily a good reason. We had accessibility issues with a popular React datepicker that lead to us making improvements to the one here, porting it to fundamental-react, and using that instead, for example. Accessibility isn't always their number one concern. There are several UI5 components with accessibility issues as well.

And yes, React has tools that can help add better keyboard functionality but it is still a manual process to implement it; it doesn't come for free. It's unfortunate that the spec didn't take into consideration the native html markup that is built to support this and this design can't easily be implemented with the correct markup. 

If it's not possible to use inputs, the second best alternative is to ensure the documentation explicitly states the expected behavior for voice readouts and keyboard handling.

Comment (IlyaSurmay) [reply stories/slider/slider.stories.js:21]: Implementation libraries would still have to implement keyboard functionality regardless of what's used,`div` or `input` because spec describes such functionality as `jump` (incrementing/decrementing slider's value by a bigger step when `shift + arrow key` is pressed) which is something that native `input` doesn't provide.

Slider is an easier component to implement than a datepicker and I don't think adding a11y functionality will be a big problem (because of the fact that a lot of other libraries is using this approach and their implementation of a11y is pretty much the same which makes me think a11y isn't that hard to implement as it may seems)

Also I found W3 ARIA examples for sliders (both single and range) and they're using `div` with no mention that `input[type="range"]` is more preferable. 
https://www.w3.org/TR/wai-aria-practices-1.1/examples/slider/slider-1.html
https://www.w3.org/TR/wai-aria-practices-1.1/examples/slider/multithumb-slider.html

I added a section in the docs that describes what attributes need to be added in order to make the slider accessible.

image

**Review by prsdthkr — APPROVED** [review]
QA LGTM after accessibility callouts.

**Review by meganaconley — APPROVED** [review]

**Pull request merged**
